docs/design/sql/14-stream-to-stream-join.md

```diff
@@ -0,0 +1,236 @@
+- stream to stream, both data sources can be streaming (e.g. messaging topics) that need to be joined.
+
+#### Goals
+
+Hazelcast already supports static to stream joins. This work is intended to introduce stream to stream joins. The main
+goals are:
+
+- SQL based / Non-Java friendly: easy to use format also for non-Java developers to use stream to stream joins
+- different join types (`INNER` , `LEFT`/`RIGHT`/`FULL` `OUTER` `JOIN`s.
+
+It should be possible to efficiently manage state for these joins using watermarks or other mechanism(s).
+
+### Functional Design
+
+#### Summary of Functionality
+
+User may execute stream-to-stream `JOIN` SQL query in the similar manner as usual JOIN. Let's define `orders`
+and `deliveries` stream events (note, `Stream` type does not exist, used just for example) :
+
+```sql
+CREATE MAPPING orders (order_id BIGINT, order_time TIMESTAMP, item_id BIGINT) TYPE Stream
+CREATE MAPPING deliveries (delivery_id BIGINT, order_id BIGINT, delivery_time TIMESTAMP, address VARCHAR) TYPE Stream
+```
+
+Consider the following query:
+
+```sql
+ CREATE VIEW fast_delivered_orders AS
+ SELECT * FROM IMPOSE_ORDER(TABLE(orders), DESCRIPTOR(order_time), INTERVAL '1' MINUTE) AS o
+ JOIN SELECT * FROM IMPOSE_ORDER(TABLE(deliveries), DESCRIPTOR(order_time), INTERVAL '1' MINUTE) AS d
+ ON d.delivery_time BETWEEN o.order_time AND o.order_time + INTERVAL '1' HOUR
+ 
+> SELECT * FROM fast_delivered_orders
+```
+
+Expected result set is represented in Table 1:
+
+| order_id | order_time       | item_id | delivery_id | order_id | delivery_time    | address     |
+|----------|------------------|---------|-------------|----------|------------------|-------------|
+| 1        | 01.03.2022:10:00 | 100     | 1           | 1        | 01.03.2022:10:40 | 'address'   |
+| 2        | 01.03.2022:11:00 | 100     | 2           | 2        | 01.03.2022:11:20 | 'address_2' |
+| ---      | ---              | ---     | ---         | ---      | ---              | ---         |
+
+__Table 1__
+
+### Technical Design
+
+_This is the technical portion of the design document. Explain the design in sufficient detail._
+
+SQL engine should use a specialized Jet processor to perform JOIN operation for two input stream events.
+
+#### Overall design
+
+Consider having two streams - __S1__ and __S2__.
+
+Let's describe all possible scenarios for joining these two streams:
+
+| Behavior                                                   | Outcome                           |
+|------------------------------------------------------------|-----------------------------------|
+| Both __S1__ **and** __S2__ don't have watermarks           | exception during query processing |
+| At least one input stream doesn't have watermarks          | exception during query processing |
+| Both __S1__ and __S2__ have watermarks with same lag times | Example in Table 3                |
+| __S1__ has zero lag and __S2__ has non-zero lag            | Example in Table 4                |
```

Comment:
Do we need to have a special case for zero lag? I think the processor doesn't care at all about any lag, it uses only the input watermarks in whatever way they are generated.

Reply:
I think it helps understanding how this works in general, so I would vote for leaving this for reference.